src/Elastic.Apm/Helpers/SystemInfoHelper.cs

```diff
@@ -0,0 +1,106 @@
+using System;
+using System.IO;
+using System.Text.RegularExpressions;
+using Elastic.Apm.Api;
+using Elastic.Apm.Logging;
+
+namespace Elastic.Apm.Helpers
+{
+	internal class SystemInfoHelper
+	{
+		private const string ContainerUidRegex = "^[0-9a-fA-F]{64}$";
+
+		private const string PodRegex = "(?:^/kubepods/[^/]+/pod([^/]+)$)|"
+			+ "(?:^/kubepods\\.slice/kubepods-[^/]+\\.slice/kubepods-[^/]+-pod([^/]+)\\.slice$)";
+
+		private static Api.System ParseContainerId(string line)
+		{
+			//Copied from the Java agent, and C#-ified
+
+			string kubernetesPodUid = null;
+
+			var fields = line.Split(':');
+			if (fields.Length != 3) return null;
+
+			var cGroupPath = fields[2];
+
+
+			var idPart = Path.GetFileName(cGroupPath);
+
+			if (string.IsNullOrWhiteSpace(idPart)) return null;
+
+			// Legacy, e.g.: /system.slice/docker-<CID>.scope
+			if (idPart.EndsWith(".scope"))
+				idPart = idPart.Substring(0, idPart.Length - ".scope".Length).Substring(idPart.IndexOf("-", StringComparison.Ordinal) + 1);
+
+			// Looking for kubernetes info
+			var dirPathPart = Directory.GetParent(cGroupPath); // Path.get (cGroupPath);  cGroupPath Path.GetPathRoot(cGroupPath);
+			if (dirPathPart != null)
+			{
+				var dir = dirPathPart.ToString();
+
+				var pattern = new Regex(PodRegex);
+				var matcher = pattern.Match(dir);
+
+				if (matcher.Success)
+				{
+					for (var i = 1; i <= matcher.Groups.Count; i++)
+					{
+						var podUid = matcher.Groups[i].Value;
+						if (string.IsNullOrWhiteSpace(podUid)) continue;
+
+						//logger.debug("Found Kubernetes pod UID: {}", podUid);
+						// By default, Kubernetes will set the hostname of the pod containers to the pod name. Users that override
+						// the name should use the Downward API to override the pod name.
```

Comment:
What is "Downward API"?

Reply:
It's [this thing](https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/#the-downward-api) - I don't know anything about it either... More Kubernetes things will come, this PR is just about the container ID, but I thought I leave the comments there, since they will be probably very useful hints once we start working on reporting Kubernetes related things.